hazelcast/src/main/java/com/hazelcast/query/impl/getters/CompactGetter.java

```diff
@@ -30,18 +30,33 @@ public CompactGetter(InternalSerializationService serializationService) {
         this.serializationService = serializationService;
     }
 
-    @Override
-    public Object getValue(Object target, String fieldPath) throws Exception {
+    private Object getValueInternal(Object target, String fieldPath, boolean useLazyDeserialization) throws Exception {
         InternalGenericRecord record;
         if (target instanceof CompactGenericRecord) {
             record = (InternalGenericRecord) target;
         } else {
             record = serializationService.readAsInternalGenericRecord((Data) target);
         }
-        GenericRecordQueryReader reader = new GenericRecordQueryReader(record);
+        GenericRecordQueryReader reader = new GenericRecordQueryReader(record, useLazyDeserialization);
         return reader.read(fieldPath);
     }
 
+    @Override
+    public Object getValue(Object target, String fieldPath) throws Exception {
+        return getValueInternal(target, fieldPath, false);
+    }
+
+    @Override
+    Object getValue(Object obj, String attributePath, Object metadata) throws Exception {
+        if (metadata == null) {
+            return getValue(obj, attributePath);
+        }
+        int depth = (int) metadata;
+        // Use lazy deserialization if we use a field of the nested field or a deeper access.
+        boolean useLazyDeserialization = depth > 1;
+        return getValueInternal(obj, attributePath, useLazyDeserialization);
```

Comment:
But what if I do `SELECT (foo).bar.baz FROM xxx`, and `baz` is also a compact type? Now, if I see it correctly, you will return an internal generic record, but it should have been a deserialized generic record

Reply:
Added a test about it to CompactNestedFieldsTest. It returns a deserialized generic record. I removed the depth logic as well. I think we don't need it. ColumnExpression will return DeserializedGenericRecord if a compact field is accessed because it is the actual column access operation. FieldAccessOperation will always return the lazy internal generic record.